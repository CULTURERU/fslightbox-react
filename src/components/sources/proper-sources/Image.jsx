import React from 'react';
import { SOURCE_CLASS_NAME } from "../../../constants/classes-names";

const Image = (
    {
        fsLightbox: {
            collections: { sourceLoadHandlers },
            elements: { sources: sourcesElements },
            props: { customAttributes, sources }
        }, i
    }
) => (
    <img
        onLoad={sourceLoadHandlers[i].handleImageLoad}
        className={SOURCE_CLASS_NAME}
        ref={sourcesElements[i]}
        src={sources[i]}
        {...(customAttributes && customAttributes[i] ? customAttributes[i] : {})}
    />
);

export default Image;
