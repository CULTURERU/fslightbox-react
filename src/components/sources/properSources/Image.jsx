import React, { Component } from 'react';
import PropTypes from "prop-types";

class Image extends Component {

    constructor(props) {
        super(props);
        this.imageOnLoad = this.imageOnLoad.bind(this);
    }

    imageOnLoad(e) {
        if (this.props._.sourcesData.isSourceAlreadyLoadedArray[this.props.i]) {
            return;
        }

        this.props._.sourcesData.sourcesDimensions[this.props.i] = {
            width: e.target.width,
            height: e.target.height
        };
        this.props.onFirstSourceLoad();
    }

    render() {
        return (
            <>
                <img
                    onLoad={ this.imageOnLoad }
                    className={ "fslightbox-single-source" }
                    ref={ this.props._.elements.sources[this.props.i] }
                    src={ this.props._.data.urls[this.props.i] }
                    alt={ this.props._.data.urls[this.props.i] }
                />
            </>
        );
    }
}


Image.propTypes = {
    _: PropTypes.object.isRequired,
    i: PropTypes.number.isRequired,
    onFirstSourceLoad: PropTypes.func.isRequired,
};
export default Image;