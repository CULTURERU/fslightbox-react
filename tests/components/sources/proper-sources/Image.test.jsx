import React from 'react';
import Image from "../../../../src/components/sources/proper-sources/Image";
import { mount, shallow } from "enzyme";

const sourceController = {
    setIndex: () => {},
    setSourceWidth: () => {},
    setSourceHeight: () => {},
    normalLoad: () => {},
    initialLoad: () => {}
};
let fsLightbox = {
    data: {
        urls: [],
    },
    sourcesData: {
        isSourceAlreadyInitializedArray: []
    },
    elements: {
        sources: [{
            current: null
        }]
    },
    core: {
        sourceController: sourceController
    }

};

const onLoadEvent = {
    target: {
        width: 0,
        height: 0
    }
};
let image;

describe('ref to sources array in fsLightbox object', () => {
    beforeAll(() => {
        image = mount(<Image
            fsLightbox={ fsLightbox }
            index={ 0 }
        />);
    });

    it('should be equal to image', () => {
        expect(fsLightbox.elements.sources[0].current).toEqual(image.getDOMNode());
    });
});

describe('onload', () => {
    describe('initial load', () => {
        beforeAll(() => {
            sourceController.setIndex = jest.fn();
            sourceController.setSourceHeight = jest.fn();
            sourceController.setSourceWidth = jest.fn();
            sourceController.normalLoad = jest.fn();
            sourceController.initialLoad = jest.fn();
            // source is loaded for first time
            fsLightbox.sourcesData.isSourceAlreadyInitializedArray[0] = false;
            image = shallow(<Image
                fsLightbox={ fsLightbox }
                index={ 0 }
            />);
            onLoadEvent.target.width = 100;
            onLoadEvent.target.height = 150;
            image.simulate('load', onLoadEvent);
        });

        it('should not call normalLoad', () => {
            expect(sourceController.normalLoad).not.toBeCalled();
        });

        it('should call setIndex with 0', () => {
            expect(sourceController.setIndex).toBeCalledWith(0);
        });

        it('should call setSourceWidth with 100', () => {
            expect(sourceController.setSourceWidth).toBeCalledWith(100);
        });

        it('should call setSourceHeight with 150', () => {
            expect(sourceController.setSourceHeight).toBeCalledWith(150);
        });

        it('should call initialLoad', () => {
            expect(sourceController.initialLoad).toBeCalled();
        });
    });


    describe('normal load', () => {
        beforeAll(() => {
            sourceController.setIndex = jest.fn();
            sourceController.setSourceHeight = jest.fn();
            sourceController.setSourceWidth = jest.fn();
            sourceController.normalLoad = jest.fn();
            sourceController.initialLoad = jest.fn();
            // source is loaded for first time
            fsLightbox.sourcesData.isSourceAlreadyInitializedArray[0] = true;
            image = shallow(<Image
                fsLightbox={ fsLightbox }
                index={ 0 }
            />);
            image.simulate('load', onLoadEvent);
        });

        it('should not call setSourceWidth', () => {
            expect(sourceController.setSourceWidth).not.toBeCalled();
        });

        it('should not call setSourceHeight', () => {
            expect(sourceController.setSourceHeight).not.toBeCalled();
        });

        it('should not call initialLoad', () => {
            expect(sourceController.initialLoad).not.toBeCalled();
        });

        it('should call setIndex with 0', () => {
            expect(sourceController.setIndex).toBeCalledWith(0);
        });

        it('should call normalLoad', () => {
            expect(sourceController.normalLoad).toBeCalled();
        });
    });
});

describe('Image DOM', () => {
    beforeAll(() => {
        image = shallow(<Image fsLightbox={ fsLightbox } index={ 0 }/>);
    });

    it('should match snapshot', () => {
        expect(image).toMatchSnapshot();
    });
});

