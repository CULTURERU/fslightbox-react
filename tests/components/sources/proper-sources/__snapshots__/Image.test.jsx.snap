// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image DOM should match snapshot 1`] = `
<img
  className="fslightbox-source fslightbox-opacity-0"
  onLoad={[Function]}
/>
`;
